### ISSUE TITLE
Longer Inference Time Compared to OSEDiff

### ISSUE THREAD

**Issue by ZHT150798** [ArcticHare105/S3Diff#20]
Hello, first of all, thank you for your prompt response to my previous question!
I noticed that despite both being single-step generation methods, OSEDiff is approximately 5 times faster than S3Diff (at the cost of fidelity). I’m curious about the reasons behind the slower inference speed — could it be related to the DE network? Additionally, I tried setting the LoRA rank of OSEDiff to the same value as mentioned in your paper, but the results didn’t improve significantly. What do you think could be the possible reasons for this?

**Comment by ch3nboyu** [comment]
S3Diff:
- Estimating the Deg_score for each image to guide LoRA.
- Additional time consumption is incurred by running the UNet twice for pos_prompt and neg_prompt, among other steps.
- more
These are very time-consuming.

you can
- Fixed Deg_score.
- ```
      lq_latent = torch.cat([lq_latent] * 2)
      encoder_hidden_states = torch.cat([neg_caption_enc, pos_caption_enc])
      model_out = self.unet(lq_latent, self.timesteps, encoder_hidden_states=encoder_hidden_states).sample
      neg_model_pred, pos_model_pred = model_out.chunk(2)
      model_pred = neg_model_pred + self.guidance_scale * (pos_model_pred - neg_model_pred)

**Comment by ZHT150798** [comment]
I used the time library to measure the execution time and found that, for the same input, the single execution of the step:
self.unet(lq_latent, self.timesteps, encoder_hidden_states=encoder_hidden_states).sample
takes significantly less time in OSEDiff compared to S3Diff. What do you think could be the possible reasons for this?

**Comment by ch3nboyu** [comment]
Hello, the possible reasons I think:

- The base model of OSED is `SD-2-1-base`, while S3 is `SD-turbo`. There may be performance differences between two base models.
- S3 uses a degradation-guided LoRA, injecting new layers into the VAE and UNet and rewriting the forward function. This could also be a reason for the difference in their time consumption. You can see the `.de_mod` in `S3Diff/src/s3diff_tile.py 269-294` and the rewritten parts by the author in `S3Diff/src/model.py 46-55`.

```
...
module.de_mod = vae_embed.reshape(-1, self.lora_rank_vae, self.lora_rank_vae)
...
module.de_mod = unet_embed.reshape(-1, self.lora_rank_unet, self.lora_rank_unet)
```

```
if not self.use_dora[active_adapter]:
    _tmp = lora_A(dropout(x))
    if isinstance(lora_A, torch.nn.Conv2d):
        _tmp = torch.einsum('...khw,...kr->...rhw', _tmp, self.de_mod)
    elif isinstance(lora_A, torch.nn.Linear):
        _tmp = torch.einsum('...lk,...kr->...lr', _tmp, self.de_mod)
    else:
        raise NotImplementedError('only conv and linear are supported yet.')

    result = result + lora_B(_tmp) * scaling
```

There may also be other reasons for such a significant difference

**Comment by ZHT150798** [comment]
> S3Diff:
> 
> * Estimating the Deg_score for each image to guide LoRA.
> * Additional time consumption is incurred by running the UNet twice for pos_prompt and neg_prompt, among other steps.
> * more
>   These are very time-consuming.
> 
> you can
> 
> * Fixed Deg_score.
> * ```
>       lq_latent = torch.cat([lq_latent] * 2)
>       encoder_hidden_states = torch.cat([neg_caption_enc, pos_caption_enc])
>       model_out = self.unet(lq_latent, self.timesteps, encoder_hidden_states=encoder_hidden_states).sample
>       neg_model_pred, pos_model_pred = model_out.chunk(2)
>       model_pred = neg_model_pred + self.guidance_scale * (pos_model_pred - neg_model_pred)
>   ```


Unfortunately, the last segment of code ended up increasing the inference time.

**Comment by ZHT150798** [comment]
@ch3nboyu 
Thank you for your response; however, unfortunately, this code ended up increasing the inference time.
> ```
>       lq_latent = torch.cat([lq_latent] * 2)
>       encoder_hidden_states = torch.cat([neg_caption_enc, pos_caption_enc])
>       model_out = self.unet(lq_latent, self.timesteps, encoder_hidden_states=encoder_hidden_states).sample
>       neg_model_pred, pos_model_pred = model_out.chunk(2)
>       model_pred = neg_model_pred + self.guidance_scale * (pos_model_pred - neg_model_pred)
>   ```

**Comment by ArcticHare105** [comment]
> Hello, first of all, thank you for your prompt response to my previous question! I noticed that despite both being single-step generation methods, OSEDiff is approximately 5 times faster than S3Diff (at the cost of fidelity). I’m curious about the reasons behind the slower inference speed — could it be related to the DE network? Additionally, I tried setting the LoRA rank of OSEDiff to the same value as mentioned in your paper, but the results didn’t improve significantly. What do you think could be the possible reasons for this?

Thank you for your question. As @ch3nboyu mentioned, we use the degradation-aware LoRA. This method cannot be merged directly into the model because it requires calculating the degradation-aware matrix for each sample and injecting it into the model dynamically. As a result, the inference time is longer compared to OSEDiff.

However, you can try fixing the degradation scores to (0.5, 0.5) for noise and blur. In this configuration, our method still achieves better results than OSEDiff. By using fixed degradation scores, we can precompute the degradation-aware matrix and merge the LoRAs into the model, which significantly reduces inference time.
